File: Section12-15 Algorithms/Section13 Sorting/173 Merge Sort.js
const numbers = [99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0];

function mergeSort (array) {
  if (array.length === 1) {
    return array
  }
  // Split Array in into right and left
  const midIndex = array.length/2;
  const left = array.slice(0, midIndex);
  const right = array.slice(midIndex);
  // console.log('left:', left);
  // console.log('right:', right);

  return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right){
  // console.log();
  // console.log('left:', left);
  // console.log('right:', right);

  if(null === left){
    return right;
  }

  if(null === right){
    return left;
  }

  let indexLeft = 0;
  let indexRight = 0;
  const newArray = [];

  while(indexLeft<left.length && indexRight<right.length){
    if(left[indexLeft] < right[indexRight]){
      newArray.push(left[indexLeft]);
      indexLeft++;
    }else{
      newArray.push(right[indexRight]);
      indexRight++;
    }
  }
  // console.log("newArray: ", newArray.concat(left.slice(indexLeft)).concat(right.slice(indexRight)));
  return newArray.concat(left.slice(indexLeft)).concat(right.slice(indexRight));
}


const answer = mergeSort(numbers);
console.log(answer);

File: Section12-15 Algorithms/Section13 Sorting/173 Merge Sort.java
class Main {
  public static void main(String[] args) {

		int[] numbers = {99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0};

	    int[] result = mergeSort(numbers);

		for(int i: result) {
			System.out.print(" " + i);
		}
	}

	public static int[] mergeSort(int[] array) {
		if(array.length < 2) return array;

		int length = array.length;
		int midIndex = length/2;
	    int[] left = Arrays.copyOfRange(array, 0, midIndex);
	    int[] right = Arrays.copyOfRange(array, midIndex, length);

	    return merge(mergeSort(left), mergeSort(right));
	}

	public static int[] merge(int[] left, int[] right) {
		int[] result = new int[left.length + right.length];
		int indexLeft = 0;
		int indexRight = 0;
		int index = 0;

		while(indexLeft<left.length && indexRight<right.length) {
			if(left[indexLeft] < right[indexRight]){
				result[index] = left[indexLeft];
				indexLeft++;
			}else{
				result[index] = right[indexRight];
				indexRight++;
			}

			index++;
		}

		//copy the remaining elements of left[], if there are any
		while(indexLeft < left.length) {
			result[index] = left[indexLeft];
			indexLeft++;
			index++;
		}

	    //Copy the remaining elements of right[], if there are any
		while(indexRight < right.length) {
			result[index] = right[indexRight];
			indexRight++;
			index++;
		}

		return result;
	}
}
